06x-new-kafka-roller.md

```diff
@@ -0,0 +1,326 @@
+
+4. **Categorize Nodes:**
+   Group nodes based on their state and connectivity:
+   - `RESTART_NOT_RUNNING`: Nodes in `NOT_READY` state.
+   - `WAIT_FOR_LOG_RECOVERY`: Nodes in `RECOVERING` state.
+   - `RESTART_UNRESPONSIVE`: Nodes unresponsive via Admin API.
+   - `MAYBE_RECONFIGURE_OR_RESTART`: Broker nodes with empty reason lists and no previous restarts/reconfigurations.
+   - `RESTART`: Nodes with reasons for restart and no previous restarts.
+   - `NOP`: Nodes needing no operation.
+
+5. **Wait for Log Recovery:**
+   Wait for `WAIT_FOR_LOG_RECOVERY` nodes to become `READY` within `operationTimeoutMs`. If timeout is reached and `numRetries` exceeds `maxRetries`, throw `UnrestartableNodesException`. Otherwise, increment `numRetries` and repeat from step 2.
+
+6. **Restart `RESTART_NOT_RUNNING` Nodes:**
+   Restart nodes in `NOT_RUNNING` state, considering special conditions:
+   - If all controller nodes are `NOT_RUNNING`, restart them in parallel to form a quorum.
+      > This is to address the issue described in https://github.com/strimzi/strimzi-kafka-operator/issues/9426.
+   - Restart `NOT_RUNNING` nodes with `POD_HAS_OLD_REVISION` in parallel. This is because, if the node is not running at all, then restarting it likely won't make any difference unless the node is out of date.
+      > For example, if a pod is in pending state due to misconfigured affinity rule, there is no point restarting this pod again or restarting other pods, because that would leave them in pending state as well. If the user then fixed the misconfigured affinity rule, then we should detect that the pod has an old revision, therefore should restart it so that pod is scheduled correctly and runs.
+   - Wait for each node's state to transition to `READY` within `operationTimeoutMs`. If timeout is reached, increment `numRetries` and repeat from step 2.
+
+7. **Restart `RESTART_UNRESPONSIVE` Nodes:**
+   Restart unresponsive nodes one by one in the order: pure controller, mixed, and broker nodes. Wait for each node's state to transition to `READY` within `operationTimeoutMs`. If timeout is reached, increment `numRetries` and repeat from step 2.
+
+8. **Refine `MAYBE_RECONFIGURE_OR_RESTART` Nodes:**
+   Describe Kafka configurations via Admin API:
+   - Nodes with dynamic config changes are added to `RECONFIGURE` group.
+   - Nodes with non dynamic config changes are added `RESTART` group.
+   - Nodes with no config changes are added to `NOP` group.
+
+9. **Reconfigure Nodes:**
+   Reconfigure nodes in the `RECONFIGURE` group:
+   - Check if `numReconfigAttempts` exceeds `maxReconfigAttempts`. If exceeded, add a restart reason and repeat from step 2. Otherwise, continue.
+   - Send `incrementalAlterConfig` request, transition state to `UNKNOWN`, and increment `numReconfigAttempts`.
+   - Wait for each node's state to transition to `READY` within `operationTimeoutMs`. If timeout is reached, repeat from step 2, otherwise continue.
```

Comment:
By definition the node's state will already be READY (otherwise it would have been in the `RESTART_NOT_RUNNING` group), therefore there is no transition to observe. 

It's never been clear to me what would be a good safety check on a dynamic reconfig. Some of the reconfigurable configs could easily result in a borked cluster, so it feels like some kind of check is needed. I think we need to take into account any effects of the reconfiguring of this node on the other nodes in the cluster. I guess step 10 is intended to achieve this, but it's not clear to me how step 10 different from just always restarting from step 2 after each reconfiguration.

Reply:
The roller will transition the node to UNKNOWN state after taking an action so that the state can be observed again, but you are right, that would likely to return READY immediately. As you said, the step 10 will repeat from step 2 if at the point the reconfigured node has gone bad. When repeating from step 2, if the node is not ready but since there is no reason to restart and reconfigure (because it's already been reconfigured) , we would end up waiting for it to become ready until the reconciliation fails.  Perhaps, we could fail the reconciliation with an error indicating that a node is not ready after reconfiguration, so we notify the human operator to do the investigation through the log.